File: source_repo/src/parser/codeblocks.js
'use strict';

/**
 * Extract fenced code blocks from Markdown body content.
 *
 * Supports both backtick (```) and tilde (~~~) fencing.
 *
 * Each code block includes:
 *  - language: the language identifier (empty string if none)
 *  - code: the code content (without fences)
 *  - startLine: 1-based line number of the opening fence
 *  - endLine: 1-based line number of the closing fence
 *
 * @param {string} body
 * @returns {Array<{language: string, code: string, startLine: number, endLine: number}>}
 */
function extractCodeBlocks(body) {
  const blocks = [];
  const lines = body.split('\n');
  let currentBlock = null;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const trimmed = line.trimStart();

    if (currentBlock === null) {
      // Check for opening fence
      const openMatch = trimmed.match(/^(`{3,}|~{3,})(\S*)$/);
      if (openMatch) {
        currentBlock = {
          fence: openMatch[1][0],          // '`' or '~'
          fenceLen: openMatch[1].length,
          language: openMatch[2] || '',
          codeLines: [],
          startLine: i + 1,
        };
      }
    } else {
      // Check for closing fence (must match same char and at least same length)
      const closePattern = new RegExp(
        `^\\${currentBlock.fence}{${currentBlock.fenceLen},}\\s*$`
      );

      if (closePattern.test(trimmed)) {
        blocks.push({
          language: currentBlock.language,
          code: currentBlock.codeLines.join('\n'),
          startLine: currentBlock.startLine,
          endLine: i + 1,
        });
        currentBlock = null;
      } else {
        currentBlock.codeLines.push(line);
      }
    }
  }

  // Unclosed code blocks are discarded (match common Markdown parser behavior)
  return blocks;
}

module.exports = { extractCodeBlocks };

File: golden_repo/mdforge/parser/codeblocks.py
"""Fenced code block extraction."""

import re


def extract_code_blocks(body: str) -> list[dict]:
    """Extract fenced code blocks from Markdown body content.

    Supports both backtick (```) and tilde (~~~) fencing.

    Each code block includes:
        - language: the language identifier (empty string if none)
        - code: the code content (without fences)
        - startLine: 1-based line number of the opening fence
        - endLine: 1-based line number of the closing fence

    Unclosed code blocks are discarded.
    """
    blocks = []
    lines = body.split("\n")
    current_block = None

    open_pattern = re.compile(r"^(`{3,}|~{3,})(\S*)$")

    for i, line in enumerate(lines):
        trimmed = line.lstrip()

        if current_block is None:
            # Check for opening fence
            open_match = open_pattern.match(trimmed)
            if open_match:
                current_block = {
                    "fence": open_match.group(1)[0],      # '`' or '~'
                    "fence_len": len(open_match.group(1)),
                    "language": open_match.group(2) or "",
                    "code_lines": [],
                    "start_line": i + 1,
                }
        else:
            # Check for closing fence (must match same char and at least same length)
            fence_char = current_block["fence"]
            fence_len = current_block["fence_len"]

            # Build pattern: 3+ backticks or tildes followed by optional whitespace
            if fence_char == "`":
                close_pattern = re.compile(rf"^`{{{fence_len},}}\s*$")
            else:
                close_pattern = re.compile(rf"^~{{{fence_len},}}\s*$")

            if close_pattern.match(trimmed):
                blocks.append({
                    "language": current_block["language"],
                    "code": "\n".join(current_block["code_lines"]),
                    "startLine": current_block["start_line"],
                    "endLine": i + 1,
                })
                current_block = None
            else:
                current_block["code_lines"].append(line)

    # Unclosed code blocks are discarded (match common Markdown parser behavior)
    return blocks